### ISSUE TITLE
Zone Destroy Fails When a Custom Hostname Fallback Origin is Set.

### ISSUE THREAD

**Issue by spher0boom** [cloudflare/terraform-provider-cloudflare#1474]
### Confirmation

- [X] My issue isn't already found on the issue tracker.
- [X] I have replicated my issue using the latest version of the provider and it is still present.

### Terraform and Cloudflare provider version

Terraform v1.1.6
on windows_amd64
+ provider registry.terraform.io/cloudflare/cloudflare v3.9.1

Tested with Provider 3.9.1, Terraform 1.0.7, 1.0.11, 1.1.6 and Provider 2.27.0, Terraform 1.0.7.

### Affected resource(s)

cloudflare_custom_hostname_fallback_origin

### Terraform configuration files

```hcl
resource "cloudflare_custom_hostname_fallback_origin" "app_fallback_origin" {
     zone_id  = local.app_zone_id
     origin   = local.app_zone_name
 }
```

### Debug output

[debug.txt](https://github.com/cloudflare/terraform-provider-cloudflare/files/8100281/debug.txt)

### Expected output

The cloudflare_custom_hostname_fallback_origin to be removed and the zone to be removed.

### Actual output

Error: error deleting Cloudflare Record: HTTP status 400: Deletion of this record is not allowed. This has been configured as a fallback origin server for SSL for SaaS. Link to documentation is here https://api.cloudflare.com/#custom-hostname-fallback-origin-for-a-zone-update-fallback-origin-for-custom-hostnames. (1039)

A zone created with the custom hostname fallback origin set cannot be destroyed. 

When a destroy command is run all assets are deleted, with the exception of the zone and the record that is designated the fallback record. 

Terraform produces the error above.

The fallback record is actually removed/unset when viewed via the GUI. 

With provider 2.27.0 a second run will delete the zone as expected.

With provider 3.9.1 terraform will continue to produce the error listed above, no matter how many additional destroy runs are executed.

The Console output is included below. This is from the second run, where only two resources are left.
```

module.cf_zone.module.cf_zone.cloudflare_zone.zone: Refreshing state... [id=redacted]
module.cf_zone.module.cf_custom_zone_dns[0].cloudflare_record.dns_records["a.test.redacted"]: Refreshing state... [id=redacted]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  - destroy

Terraform will perform the following actions:

  # module.cf_zone.module.cf_custom_zone_dns[0].cloudflare_record.dns_records["a.test.redacted"] will be destroyed
  - resource "cloudflare_record" "dns_records" {
      - allow_overwrite = false -> null
      - created_on      = "2022-02-18T19:48:18.070461Z" -> null
      - hostname        = "redacted" -> null
      - id              = "redacted" -> null
      - metadata        = {
          - "auto_added"             = "false"
          - "fallback_origin"        = "true"
          - "managed_by_apps"        = "false"
          - "managed_by_argo_tunnel" = "false"
          - "source"                 = "primary"
        } -> null
      - modified_on     = "2022-02-18T19:48:54.843706Z" -> null
      - name            = "redacted" -> null
      - proxiable       = true -> null
      - proxied         = true -> null
      - ttl             = 1 -> null
      - type            = "CNAME" -> null
      - value           = "www.cloudflare.com" -> null
      - zone_id         = "redacted" -> null
    }

  # module.cf_zone.module.cf_zone.cloudflare_zone.zone will be destroyed
  - resource "cloudflare_zone" "zone" {
      - id                  = "redacted" -> null
      - meta                = {
          - "phishing_detected"  = false
          - "wildcard_proxiable" = true
        } -> null
      - name_servers        = [
          - "reza.ns.cloudflare.com",
          - "seamus.ns.cloudflare.com",
        ] -> null
      - paused              = false -> null
      - plan                = "enterprise" -> null
      - status              = "active" -> null
      - type                = "full" -> null
      - vanity_name_servers = [] -> null
      - zone                = "redacted" -> null
    }

Plan: 0 to add, 0 to change, 2 to destroy.

Do you really want to destroy all resources in workspace "a"?
  Terraform will destroy all your managed infrastructure, as shown above.
  There is no undo. Only 'yes' will be accepted to confirm.

  Enter a value: yes

module.cf_zone.module.cf_custom_zone_dns[0].cloudflare_record.dns_records["a.test.redacted"]: Destroying... [id=redacted]
╷
│ Error: error deleting Cloudflare Record: HTTP status 400: Deletion of this record is not allowed. This has been configured as a fallback origin server for SSL for SaaS. Link to documentation is here https://api.cloudflare.com/#custom-hostname-fallback-origin-for-a-zone-update-fallback-origin-for-custom-hostnames. (1039)   
│
```

### Steps to reproduce

Add a custom fallback origin to a zone. Make sure the origin shows as 'active' in the GUI. A non-active/pending fallback origin will not reproduce the error.

Destroy the zone.

Get the error.

**Comment by jacobbednarz** [comment]
If I understand this correctly, this is expected behaviour. Looking at the debug logs, you can see there is a `DELETE` request issued for a DNS record that is used as a fallback origin. You cannot delete the DNS record but instead, need to delete the fallback origin configuration which in turn, allows you to remove the DNS record. This doesn't look to be related to zones but the custom hostname + DNS records using it.

**Comment by jacobbednarz** [comment]
To do this in Terraform, you _may_ need to delete them separately or include a `depends_on` to get the delete ordering correct.

**Comment by spher0boom** [comment]
I'm actually issuing a destroy command for the zone assets, not attempting to delete a single record. All other assets are properly destructed, and there aren't any dependency issues other than the fallback record.

The fallback record dependency is actually deleted (at least it appears to be deleted in the UI) with both the 2.27 and 3.9.1 providers when the destroy is issued. The record and zones are not.

- If I re-execute the destroy command under the 2.27 provider it completes the last two items (the record and zone destruction) without any errors. 

- If I re-execute the destroy command under the 3.9.1 provider it gets stuck on the same error when it attempts to delete the record again, even though the dependency setting appears to be gone in the UI. 

It seems that both providers recognize the dependency needs to be removed, and actually do it, but then don't wait long enough for it to complete before issuing the record delete. Some of the other resources issue a 'waiting for xxx' message and pause for the commands to complete. This one does not, and seems to immediately fail.

If the destroy didn't actually remove the dependency I'd feel like this was an intended behavior, but since it does actually remove the dependency, and since the 2.27 provider will actually delete the zone when re-run, doesn't this seem more like an API/Provider behavior bug than intended behavior?

Also, isn't the destroy supposed to destroy the targeted asset and all dependencies? It seems to for all of our other providers. When we destroy assets we only ever have to issue a single command, and we don't need to 'prep' the assets first.

Thanks for the quick response!

**Comment by jacobbednarz** [comment]
I suspect what is happening here is that due to the dependency tree not being recognised, it is just attempting to delete all the resources (as you would expect) however, there is a relationship between two resources that is enforced on the API side and Terraform isn't aware of it. This type of behaviour works in other providers because: 1) the functionality doesn't exist or 2) the relationship is also embedded in Terraform.

Looking at the difference between 2.27.0 and 3.9.0, only the error message has changed in the `Delete` method so I'm very surprised to hear you're seeing behaviour differences.


<details><summary>git diff v2.27.0 v3.9.0 -- cloudflare/resource_cloudflare_custom_hostname_fallback_origin.go</summary>

```diff
diff --git cloudflare/resource_cloudflare_custom_hostname_fallback_origin.go cloudflare/resource_cloudflare_custom_hostname_fallback_origin.go
index e3d09737..263d2577 100644
--- cloudflare/resource_cloudflare_custom_hostname_fallback_origin.go
+++ cloudflare/resource_cloudflare_custom_hostname_fallback_origin.go
@@ -7,13 +7,14 @@ import (
 	"strings"
 
 	"github.com/cloudflare/cloudflare-go"
-	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
-	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/pkg/errors"
 )
 
 func resourceCloudflareCustomHostnameFallbackOrigin() *schema.Resource {
 	return &schema.Resource{
+		Schema: resourceCloudflareCustomHostnameFallbackOriginSchema(),
 		Create: resourceCloudflareCustomHostnameFallbackOriginCreate,
 		Read:   resourceCloudflareCustomHostnameFallbackOriginRead,
 		Update: resourceCloudflareCustomHostnameFallbackOriginUpdate,
@@ -21,23 +22,6 @@ func resourceCloudflareCustomHostnameFallbackOrigin() *schema.Resource {
 		Importer: &schema.ResourceImporter{
 			State: resourceCloudflareCustomHostnameFallbackOriginImport,
 		},
-
-		SchemaVersion: 0,
-		Schema: map[string]*schema.Schema{
-			"zone_id": {
-				Type:     schema.TypeString,
-				ForceNew: true,
-				Required: true,
-			},
-			"origin": {
-				Type:     schema.TypeString,
-				Required: true,
-			},
-			"status": {
-				Type:     schema.TypeString,
-				Computed: true,
-			},
-		},
 	}
 }
 
@@ -47,7 +31,7 @@ func resourceCloudflareCustomHostnameFallbackOriginRead(d *schema.ResourceData,
 
 	customHostnameFallbackOrigin, err := client.CustomHostnameFallbackOrigin(context.Background(), zoneID)
 	if err != nil {
-		return errors.Wrap(err, fmt.Sprintf("error reading custom hostname fallback origin %q", zoneID))
+		return fmt.Errorf("error reading custom hostname fallback origin %q: %w", zoneID, err)
 	}
 
 	d.Set("origin", customHostnameFallbackOrigin.Origin)
@@ -62,7 +46,7 @@ func resourceCloudflareCustomHostnameFallbackOriginDelete(d *schema.ResourceData
 
 	err := client.DeleteCustomHostnameFallbackOrigin(context.Background(), zoneID)
 	if err != nil {
-		return errors.Wrap(err, "failed to delete custom hostname fallback origin")
+		return fmt.Errorf("failed to delete custom hostname fallback origin: %w", err)
 	}
 
 	return nil
@@ -80,16 +64,18 @@ func resourceCloudflareCustomHostnameFallbackOriginCreate(d *schema.ResourceData
 	return resource.Retry(d.Timeout(schema.TimeoutDefault), func() *resource.RetryError {
 		_, err := client.UpdateCustomHostnameFallbackOrigin(context.Background(), zoneID, fallbackOrigin)
 		if err != nil {
-			if err.(*cloudflare.APIRequestError).InternalErrorCodeIs(1414) {
+			//nolint:errorlint
+			if errors.As(err, &cloudflare.APIRequestError{}) && err.(*cloudflare.APIRequestError).InternalErrorCodeIs(1414) {
 				return resource.RetryableError(fmt.Errorf("expected custom hostname resource to be ready for modification but is still pending"))
+			} else {
+				return resource.NonRetryableError(fmt.Errorf("failed to create custom hostname fallback origin: %w", err))
 			}
-			return resource.NonRetryableError(errors.Wrap(err, "failed to create custom hostname fallback origin"))
 		}
 
 		fallbackHostname, err := client.CustomHostnameFallbackOrigin(context.Background(), zoneID)
 
 		if err != nil {
-			return resource.NonRetryableError(fmt.Errorf("failed to fetch custom hostname: %s", err))
+			return resource.NonRetryableError(fmt.Errorf("failed to fetch custom hostname: %w", err))
 		}
 
 		// Address an eventual consistency issue where deleting a fallback hostname
@@ -102,7 +88,8 @@ func resourceCloudflareCustomHostnameFallbackOriginCreate(d *schema.ResourceData
 		id := stringChecksum(fmt.Sprintf("%s/custom_hostnames_fallback_origin", zoneID))
 		d.SetId(id)
 
-		return resource.NonRetryableError(resourceCloudflareCustomHostnameFallbackOriginRead(d, meta))
+		resourceCloudflareCustomHostnameFallbackOriginRead(d, meta)
+		return nil
 	})
 
 }
@@ -119,13 +106,15 @@ func resourceCloudflareCustomHostnameFallbackOriginUpdate(d *schema.ResourceData
 	return resource.Retry(d.Timeout(schema.TimeoutDefault), func() *resource.RetryError {
 		_, err := client.UpdateCustomHostnameFallbackOrigin(context.Background(), zoneID, fallbackOrigin)
 		if err != nil {
-			if err.(*cloudflare.APIRequestError).InternalErrorCodeIs(1414) {
+			//nolint:errorlint
+			if errors.As(err, &cloudflare.APIRequestError{}) && err.(*cloudflare.APIRequestError).InternalErrorCodeIs(1414) {
 				return resource.RetryableError(fmt.Errorf("expected custom hostname resource to be ready for modification but is still pending"))
 			}
-			return resource.NonRetryableError(errors.Wrap(err, "failed to update custom hostname fallback origin"))
+			return resource.NonRetryableError(fmt.Errorf("failed to update custom hostname fallback origin: %w", err))
 		}
 
-		return resource.NonRetryableError(resourceCloudflareCustomHostnameFallbackOriginRead(d, meta))
+		resourceCloudflareCustomHostnameFallbackOriginRead(d, meta)
+		return nil
 	})
 }
 
```

</details>

Are you sure you're operating on individual resources and not seeing it work because you ran a different version of the delete in a previous test?

**Comment by spher0boom** [comment]
We’ll, in both cases I’m issuing a destroy on the entire terraform config, which is the same for both provider versions. I can double check and make sure it’s not a terraform version difference, maybe I didn’t track my testing correctly.

I can try the dependency as well, but the easiest solution may be to do a pre-destroy run to remove this dependency and downgrade the zone to a “free” plan zone so it can actually be deleted via the destroy, I was just hoping to avoid multiple runs on the resources.

Thanks for your help.

**Comment by spher0boom** [comment]
I tried adding a depends on, and it did not work. The error still occurs.

I also tried changing the setting in an external provider, but that didn't work either because the setting is literally removed from the zone, not just changed, and the terraform then pukes because that object doesn't exist to query.

The only way to do this seems to be running an apply to change the setting, and then running a separate destroy.

**Comment by spher0boom** [comment]
I did more digging, and it turns out setting the dependency does work - the Fallback Origin is removed from the Custom Hostnames page. However, it turns out there is currently a back end issue with the Fallback Origin DNS flag. The problem also occurs via the API and in the GUI.

In all instances the Fallback Origin is removed from the Custom Hostnames page, but the flag preventing deletion still exists on the DNS page (my bad, I didn't check the DNS records earlier). Every DNS record in the zone that is set and activated has the flag set, and it is not removed even if the fallback record is changed. I tested this on multiple zones, and each is now sitting with 4-5 records flagged as the Fallback Origin, even though none is set on the zone.

I don't know why I only found this when I switched providers. Bad timing I guess.

I've created ticket 2381139 on the enterprise support site for this issue. I'll close out this item since setting the dependency works.

Thanks for your help.
